### ISSUE TITLE
Drill Down Components Discussion

### ISSUE THREAD

**Issue by JesseSingleton** [stevegreatrex/birst-react-embed#2]
## General Background (How we use DD/Filters currently)
The way we currently 'Drill Down' to a dashboard in `playerGrid` is by passing in a property on `IProfileGridParameters` like so:
```
drillDown: {
        dashboard: 'Collection',
        page: 'React Embed Child Dashboard'
}
```

... where `dashboard` is the Collection that contains the child dashboard and `page` is the child dashboard that you wish to drill down to. (`Collection` rather than `dashboard` and `childDashboard` instead of `page` would make more sense in hindsight 👀 ).

This gets mapped to `IProfileCardGridProps` which is available on each `ProfileCard`and finally we use the following function `onClick : drillDownHandler()` to send the operation to the window:
```
const drilldownHandler = (props: IProfileCardProps) => {
	if(!props.drillDown)
		return;
        
        //Add a filter to the parent window
	const parentWindow = window.parent as IBirstParentWindow;
	parentWindow.filters = { drillDownFilter: props.playerName};

        //Drill Down to the child dashboard
	window.parent.postMessage( { operation: 'drillToDashboard', dashboard: props.drillDown.dashboard, page: props.drillDown.page }, '*' );
};
```

## General Idea (How we want to use DD/Filters in the future)
Sharoon and I feel that it would make sense for it to go on the `birst-react-embed` library in a reusable function to make life easier and as you put it  "_reduce the amount of WTF_".

The real general idea would still pass in the `collection` and `page` properties into `createConnectedGrid` 
 and then using a `applyDrillDown` function that transforms the `IBQLComponent` into an component with a `drillDown` (I'm not sure I explained that very well.) Then the drillDown is available on each component and fillters would be optional.

## General Implementation
We wrote some pseudo code to demonstrate how it could work:
```
const createConnectedGrid = (parameters: IProfileGridParameters) => {
    const bqlComponent = createQuery(parameters);
    
    if (parameters.drillDown) {
        bqlComponent = applyDrilldown(bqlComponent, parameters.drillDown);
    }
    
    const Connected = connectToBirst(bqlComponent)(ProfileCardGrid);
    return <Connected {...parameters} />;
};


function applyDrilldown<TBirstComponent>(component: TBirstComponent, drillDownParameters: IDrillDownParameters)  {
    function createDrillDownComponent<TDrillDownComponent extends TBirstComponent>(component: TBirstComponent, drillDownParameters: IDrillDownParameters){
        // Set drilldown on params.
        // return new component with drill down
    }
}
```
I hope that makes some sense. Here's our current implementation for [reference](https://github.com/EDGE10/birst-components/tree/profile-grid)

**Comment by stevegreatrex** [comment]
That makes sense, and I agree that the naming of the properties is better with `collection` and `dashboard` or `childDashboard`.

Sounds to me like we should pass `applyDrillDown` into the connected component as props by adding them to `IComponentState` here.  Something like how `dispatch` is injected into component props by `connect` in `react-redux`
https://github.com/stevegreatrex/birst-react-embed/blob/acaa537b8e886294b72d8db0b0b81b3cb330bb96/src/BirstComponent.tsx#L16


```typescript
//this library
interface ApplyDrillDown {
  (drillDown: IDrillDownParameters): void;
}

//example usage
interface IMyComponentProps {
  applyDrillDown: ApplyDrillDown
}

const MyComponent = (props: IMyComponentProps) => (
  <div onClick={() => props.applyDrillDown({ ... })}>Click me to drill down</div>
);
```

## Questions
❓ Do you think the consumer would ever want to _react_ to a drill down being applied (via a promise, for example) or can we assume that applying a drill down will always cause a reload and so we can fire and forget?
❓ Is it preferable to have `IDrillDownParameters` as a type or to instead accept the 2 parameters directly (`applyDrillDown('collection', 'dashboard')`)?  Are there likely (or are there already) any other parameters?  I'm leaning towards the parameters object as it means that you could define your available drill downs in code, for example `applyDrillDown(MyDashboard.MyCollection.myDashboard)`
❓ Does this account for applying filters?  Should we try to include that within the scope of this change or deal with that separately?  I have no idea what the signature or behaviour of that would look like so I can't really comment!


As a side note: some of the files in this repo really need a rename for clarity!  I'm looking at you `BirstComponent`...

**Comment by JesseSingleton** [comment]
Sounds like a good plan. The example code is exactly how I imagined you'd be using the `drillDown` prop. Not entirely clear on how you are proposing to implement it in `BirstComponent`.. 
> Something like how dispatch is injected into component props by connect in react-redux

... asked Sharoon about that and he wasn't too sure either 👀. But we like the example code. Here are some answers to your questions:

1. We can't imagine a situation where we'd want to drilldown without a reload. So I think it's safe to assume we're always going to be reload.
2. I think a type is preferable. As for other parameters the only other parameters that we could possibly have/want is an optional parameter `filter` to be able to apply filters. All a filter does is add a variable to the parent window, drills down to the child, and then the child looks on the parent window for the filter. If we could add this as an optional parameter on the drill down it would obscure more of the terrible Birst behavior.. 😄 
3. Filters are fairly easy as described above, maybe attempting to implement the drill down behaviour first might be preferable. But if it can be as simple as attaching it as an optional parameter, that would be great!

Yeah there are a few things I want to change on `profileGrid` and improve the example for that repo too!

**Comment by stevegreatrex** [comment]
Ok, sounds like a consensus.  Are you happy to take a crack at a PR or do you want to wait until I'm around next week?

**Comment by JesseSingleton** [comment]
Sorry for the late reply! I'm still working on sprint stuff currently, so I'm not sure I'll get to do anything significant. I certainly can have a look at it and then discuss next week! Your last week 😢 😢 !

**Comment by stevegreatrex** [comment]
Fixed by #4
